### PR TITLE
Fix flickering crumbling texture on faces with 45° normals

### LINKED ISSUE #9608: block breaking overlay flickers on campfire/sculk sensor/pointed dripstone/rail
**Minecraft Version:** 1.20.1

**Forge Version:** 47.0.34

**Description of issue:**

## Expected Behavior

https://github.com/MinecraftForge/MinecraftForge/assets/68275908/078f2813-3625-419e-a484-02c202395a7b

## Actual Behavior

https://github.com/MinecraftForge/MinecraftForge/assets/68275908/d9b9b594-307b-402e-92d3-5b0aacaa693c

### PR DESCRIPTION
**Pull request by Technici4n**
Fixes MinecraftForge/MinecraftForge#9608.

This is how the bug happens:
1. Get quad normal from the quad.
2. Multiply quad normal by normal pose matrix before adding it to the vertex consumer.
3. `SheetedDecalTextureGenerator` inverts the transform using the inverse normal pose matrix.
4. Then it tries to get the closest direction. Due to rounding errors with the two matrix multiplications, if the face normal has 45° angle it is basically a random direction that is picked (since the matrices change depending on the mouse movement).

I am not convinced by this added method in `Direction`. Should I make it an `@ApiStatus.Internal` method in `ForgeHooksClient` instead?

### CODE REVIEW

**Comment by shartte** [comment]
Is it feasible to add unit tests for the added method?

File: patches/minecraft/net/minecraft/core/Direction.java.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
+--- a/net/minecraft/core/Direction.java
++++ b/net/minecraft/core/Direction.java
+@@ -463,6 +_,22 @@
+       return direction;
+    }
+ 
++   /** NeoForge: Version of {@link #getNearest(float, float, float)} that accounts for potential rounding errors. */
++   public static Direction getNearestStable(float p_122373_, float p_122374_, float p_122375_) {
```
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/core/Direction.java.patch:8]: This should have proper parameter names, or if it really makes sense for them to match `getNearest` this should be added to the `exc` file.

**Review by pupnewfster — DISMISSED** [review]

File: patches/minecraft/net/minecraft/core/Direction.java.patch (hunk 1 above)
Comment (Technici4n) [reply patches/minecraft/net/minecraft/core/Direction.java.patch:8]: Added to the `exc` file.

**Review by marchermans — APPROVED** [review]

**Comment by marchermans** [comment]
@pupnewfster Can you approve this in the current form? Or are still changes needed?

**Comment by pupnewfster** [comment]
It looks fine to me, though I have not had a chance to actually test it.

**Review by pupnewfster — APPROVED** [review]

**Comment by Lolothepro** [comment]
Note: [MC-193656](https://bugs.mojang.com/browse/MC-193656)

**Pull request merged**